group-coordinator/src/main/java/org/apache/kafka/coordinator/group/AbstractGroup.java

```diff
@@ -0,0 +1,625 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group;
+
+import org.apache.kafka.clients.consumer.internals.ConsumerProtocol;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.errors.StaleMemberEpochException;
+import org.apache.kafka.common.errors.UnknownMemberIdException;
+import org.apache.kafka.common.message.ListGroupsResponseData;
+import org.apache.kafka.coordinator.group.assignor.SubscriptionType;
+import org.apache.kafka.coordinator.group.consumer.Assignment;
+import org.apache.kafka.coordinator.group.consumer.ConsumerGroup;
+import org.apache.kafka.coordinator.group.consumer.TopicMetadata;
+import org.apache.kafka.image.ClusterImage;
+import org.apache.kafka.image.TopicImage;
+import org.apache.kafka.image.TopicsImage;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+import org.apache.kafka.timeline.TimelineInteger;
+import org.apache.kafka.timeline.TimelineObject;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.apache.kafka.coordinator.group.assignor.SubscriptionType.HETEROGENEOUS;
+import static org.apache.kafka.coordinator.group.assignor.SubscriptionType.HOMOGENEOUS;
+
+public abstract class AbstractGroup implements Group {
```

Comment:
The name is not ideal because it gives the impression that all group types extends this. I would rather use something like `AbstractConsumerAndShareGroup`.

Reply:
I think your point is a good one, but listing the two current subclasses seems problematic when we add streams later on. I wonder if there's some other common feature that we could put in the name. I suppose they're non-classic protocol groups. I hesitate to suggest AbstractModernGroup, but I expect you see what I mean.

Reply:
The protocol behind the groups seems to be the main distinction as I see it (meaning the base protocol). Classic groups as part of the classic protocol, vs share/consumer/streams.. that all build on top of the new protocol). Other than that, the groups are quite different I would say, so hard to find a common feature. Seeing it like this, the non-classic seems like a sensible umbrella term (given that classic is the term we are standardizing for the old group and old protocol, so convenient hint to the protocol too). 

No matter the name, I would strongly suggest we add a the java doc for this class btw.

Reply:
My idea about AbstractGroup was that down the line AbstractGroup would be the one extended if we require additional groups and classic group is just an exception for now for backward compatibilty.

@lianetm I agree with java doc, will add.

Reply:
One way to address this would be to use the following structure:
* coordinator.classic.* for classic group related classes
* coordinator.modern.* (or nextgen) for common classes used by next gen groups
* coordinator.modern.consumer.* for specific classes
* coordinator.modern.share.*
* ...

Reply:
I also wonder if streams will really use this class because the assigned resources will be totally different, I suppose.

Reply:
I have renamed `AbstractGroup` to `AbstractModernGroup` and `GroupMember` to `ModernGroupMember`. Also I have created a common package named `modern` as suggested above.

This PR mostly does the refactor around adding abstract classes and in the next PR I will move `consumer` package under `modern`, also will move couple of shared classes from `consumer` to `modern` as well. Just don't to increase more diffs in the current PR with additional refactoring.

Reply:
Now that I see the use of `modern` in the code, I'm pretty comfortable with it.